wp-includes/sqlite-ast/class-wp-sqlite-information-schema-reconstructor.php

```diff
@@ -0,0 +1,427 @@
+		}
+		$definition[] = 'KEY';
+
+		// Remove the prefix from the index name if there is any. We use __ as a separator.
+		$index_name   = explode( '__', $key['name'], 2 )[1] ?? $key['name'];
+		$definition[] = $this->quote_sqlite_identifier( $index_name );
+
+		// Key columns.
+		$cols = array();
+		foreach ( $key_columns as $column ) {
+			// Get data type and length.
+			$data_type = strtolower( $data_types[ $column['name'] ] );
+			if ( 1 === preg_match( '/^(\w+)\s*\(\s*(\d+)\s*\)/', $data_type, $matches ) ) {
+				$data_type   = $matches[1]; // "varchar"
+				$data_length = min( $matches[2], $key_length_limit ); // "255"
+			}
+
+			// Apply max length if needed.
+			if (
+				str_contains( $data_type, 'char' )
+				|| str_starts_with( $data_type, 'var' )
+				|| str_ends_with( $data_type, 'text' )
+				|| str_ends_with( $data_type, 'blob' )
+			) {
+				$cols[] = sprintf(
+					'%s(%d)',
+					$this->quote_sqlite_identifier( $column['name'] ),
+					$data_length ?? $key_length_limit
+				);
+			} else {
+				$cols[] = $this->quote_sqlite_identifier( $column['name'] );
+			}
+		}
+
+		$definition[] = '(' . implode( ', ', $cols ) . ')';
+		return implode( ' ', $definition );
+	}
+
+	/**
+	 * Determine if a column has a default value.
+	 *
+	 * @param  string      $mysql_type    The MySQL data type of the column.
+	 * @param  string|null $default_value The default value of the SQLite column.
+	 * @return bool                       True if the column has a default value, false otherwise.
+	 */
+	private function column_has_default( string $mysql_type, ?string $default_value ): bool {
+		if ( null === $default_value || '' === $default_value ) {
+			return false;
+		}
+		if (
+			"''" === $default_value
+			&& in_array( strtolower( $mysql_type ), array( 'datetime', 'date', 'time', 'timestamp', 'year' ), true )
+		) {
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Get a MySQL column or index data type from legacy data types cache table.
+	 *
+	 * This method retrieves MySQL column or index data types from a special table
+	 * that was used by an old version of the SQLite driver and that is otherwise
+	 * no longer needed. This is more precise than direct inference from SQLite.
+	 *
+	 * @param  string      $table_name           The table name.
+	 * @param  string      $column_or_index_name The column or index name.
+	 * @return string|null                       The MySQL definition, or null when not found.
+	 */
+	private function get_cached_mysql_data_type( string $table_name, string $column_or_index_name ): ?string {
+		try {
+			$mysql_type = $this->driver->execute_sqlite_query(
+				'SELECT mysql_type FROM _mysql_data_types_cache WHERE `table` = ? AND column_or_index = ?',
+				array( $table_name, $column_or_index_name )
+			)->fetchColumn();
+		} catch ( PDOException $e ) {
+			if ( str_contains( $e->getMessage(), 'no such table' ) ) {
+				return null;
+			}
+			throw $e;
+		}
+		if ( str_ends_with( $mysql_type, ' KEY' ) ) {
```

Comment:
What are some values that would trigger this condition? Let's document that inline.

Reply:
:+1: Documented in https://github.com/Automattic/sqlite-database-integration/pull/42/commits/a8a56041dd6ca9e734b14ea9e79d822a081835e5.